File: Others/NOS/DP/6.背包DP/P1048.cpp
/*
P1048 [NOIP 2005 普及组] 采药
https://www.luogu.com.cn/problem/P1048
*/
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int T, M;
    cin >> T >> M;
    vector<pair<int, int>> items(M);
    for (int i = 0; i < M; i++) cin >> items[i].first >> items[i].second;
    vector<int> f(T + 1, 0);
    for (auto [t, v] : items)
        for (int j = T; j >= t; j--) f[j] = max(f[j], f[j - t] + v);
    cout << f[T] << endl;
    return 0;
}

File: Others/NOS/DP/6.背包DP/P1048.py
"""
P1048 [NOIP 2005 普及组] 采药
https://www.luogu.com.cn/problem/P1048
背包DP模板題，求極值
"""
T, M = map(int, input().split())
items = [list(map(int, input().split())) for _ in range(M)]

f = [0] * (T + 1)
for t, v in items:
    for j in range(T, t - 1, -1):
        f[j] = max(f[j], f[j - t] + v)

print(f[T])